### PR TITLE
ipq40xx: add VRX518 DSL support for FritzBox 7520/7530

### PR DESCRIPTION
**Pull request by dhewg**
This brings up DSL to a working state, it's been tested by the community for over two years now, see https://forum.openwrt.org/t/adding-support-for-vrx518-and-maybe-vrx320/55160

The initial patches are done by Martin Schiller (@sch-m), but that didn't get the DSL hardware in a usable state using the mainline kernel.
Later, Jan Hoffmann (@janh) was able to fix the drivers, added here as patches to the individual packages.
I just added some bits and pieces and cleaned everything up so it's hopefully ready for merge.

We already merged some preparatory bits in the past, these are the remaining patches. The elephant in the room is the license of the 3 firmware files. As of today, the state isn't clear, and as-is we cannot ship installable images with these firmware files. Only MaxLinear can solve that, and I opened an issue about it here: https://prplfoundationcloud.atlassian.net/browse/PCI-16
Alas, no visible progress for ~0.5 years :\

The options for that problem are:
1) Don't merge the firmware packages at all
2) Skip the firmware packages for official builds (e.g. !BUILDBOT)
3) Add runtime scripts instead of the firmware files which downloads them instead
4) Don't bundle firmware files in the downloadable image, but let users instead use imagebuilder to add them instead

I'd outright discard 3), since for most users the device will be the only one which provides inet access, which then makes downloading them impossible.
I prefer 2) so users who build their image themselves have a fully working one.
No idea about 4), but that may be doable on top of 2) but should not block 2)?

### CODE REVIEW

File: package/firmware/lantiq/dsl-vr11-firmware-xdsl/Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+define Package/dsl-vr11-firmware-xdsl-a/description
+	VRX518 / VR11 CPE VDSL and ADSL Annex A firmware
+endef
+
+define Build/Compile
+endef
+
+define Package/dsl-vr11-firmware-xdsl-a/install
+	$(INSTALL_DIR) $(1)/lib/firmware/
+	$(INSTALL_DATA) $(PKG_BUILD_DIR)/LICENSE $(1)/lib/firmware/xcpe_$(ANNEX_A_VER).LICENSE
```
Comment (dangowrt) [inline package/firmware/lantiq/dsl-vr11-firmware-xdsl/Makefile:41]: No need to store LICENSE file on the target itself. Better use `PKG_LICENSE` and `PKG_LICENSE_FILE`.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts [hunk 2]
```diff
@@ -17,7 +17,7 @@
 		led-boot = &power_green;
 		led-failsafe = &info_red;
 		led-running = &power_green;
-		led-upgrade = &info_green;
+		led-upgrade = &info_red;
```
Comment (dangowrt) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts:20]: This change looks unrelated to adding support for VRX518

File: package/firmware/lantiq/dsl-vr11-firmware-xdsl/Makefile (hunk 1 above)
Comment (dhewg) [reply package/firmware/lantiq/dsl-vr11-firmware-xdsl/Makefile]: Done, but there's no SPDX id for this one, so that's custom now
It's been a while, but IIRC the license is stored for the VR9 fw, where this came from

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts (hunk 2 above)
Comment (dhewg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts:20]: It clears the green LED for DSL status indicator, which is added in the same patch. @janh did this, I assume to match the behavior of the vendor fw

Comment (dhewg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts:20]: Actually, mostly because the LEDs are labeled on the case, which I removed :)

Comment (janh) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dts:20]: The vendor firmware uses the power LED (labeled "Power/DSL") for DSL status, and according to the manual the green info LED is used as upgrade indicator.

But that setup isn't ideal, as it makes the power LED ambiguous. And other Fritzboxes already use the green info LED for DSL status in OpenWrt, and (except for 7430) one of the red LEDs as upgrade indicator, so it seems like a better idea to match that.

**Comment by robimarko** [comment]
I am not sure if legally it should even be possible for the build-system to download the FW at all.

**Comment by dhewg** [comment]
Maybe, which is why I propose to shift that burden to the end user.

Similar to BUILD_PATENTED, which excluded e.g. patented codecs for legal reasons, we could introduce a toggle ALLOW_NON_FREE_FIRMWARE or something, which these firmware packages could then depend on.
If that toggle in return depends on !BUILDBOT, the official images would not be tainted (IIUC), and each user could decide which way to go, making it their decision and responsibility.

**Comment by robimarko** [comment]
I dont know if that would be enough for the project to distance itself, but seems like a good start.
I am not a lawyer, seems like we would need one for this kind of things.

**Comment by robimarko** [comment]
Users can add the FW to `files` dir if compiling manually or otherwise just copy over SCP.

**Comment by dhewg** [comment]
> This was discussed during the last developer meeting. We decided that if a firmware does not allow redistribution, we will not add it in OpenWrt at all, not even behind a config symbol like ALLOW_NON_FREE_FIRMWARE. See https://openwrt.org/meetings/20221130#firmware_packages_with_eula

Okay, thanks for the info! It sucks for the end user, but yeah, sucky licenses we unfortunately have to live with :\
I'll adapt the PR and dependencies accordingly soon-ish.

**Comment by dangowrt** [comment]
At least up to now it was also ok to have the user download and extract the firmware at runtime with some automated help, e.g.
https://github.com/openwrt/openwrt/blob/master/package/kernel/lantiq/ltq-vdsl-fw/src/vdsl_fw_install.sh

**Comment by dhewg** [comment]
But that's for lantiq, where we have an initial working firmware with a compatible license. That scripts upgrades to another non-compatible one, because the initial fw doesn't support vectoring, where the downloaded one does.

Here we don't have any firmware at all, downloading anything only works if the device has inet access through another device. That's even more painful that scp'ing 3 files IMHO. So I'll just drop the fw packages from the PR and add instructions instead.

**Comment by mrkiko** [comment]
On Tue, 10 Jan 2023, Robert Marko wrote:

> Date: Tue, 10 Jan 2023 14:11:34
> From: Robert Marko ***@***.***>
> Reply-To: openwrt/openwrt
>     ***@***.***>
> To: openwrt/openwrt ***@***.***>
> Cc: mrkiko ***@***.***>, Mention ***@***.***>
> Subject: Re: [openwrt/openwrt] ipq40xx: add VRX518 DSL support for FritzBox
>     7520/7530 (PR #11728)
> 
> 
> Users can add the FW to files dir if compiling manually or otherwise just copy over SCP.
> 
> @mrkiko You had some thoughts on this

Yes, I think the best solution would be the user to do the required steps manually.
Furthermore, the user is actually instructed to remove the AVM fs ubi volumes upon installation.
We might investigate how hard it could be to rework the installation instructions so that, before overwriting kernel with our u-boot and deleting the UBI AVM fs, a user could mount it somehow and extract the required firmware file(s).

**Comment by dhewg** [comment]
The firmware packages are dropped now, instead there's a TODO added to the packages which depend on a firmware.
The last commit now contains instructions how to get these.

**Comment by moeller0** [comment]
Quick note the wget commands in the last patch only pull some html document not the desired files. Following the links in a browser allows downloading them.
For ppe_fw.bin the following link seems to work:
`https://gitlab.com/prpl-foundation/intel/vrx518_ppe_fw/-/raw/ugw_8.5.2.10/platform/xrx500/ppe_fw.bin`
instead of:
`https://gitlab.com/prpl-foundation/intel/vrx518_ppe_fw/-/blob/ugw_8.5.2.10/platform/xrx500/ppe_fw.bin`
so maybe `blob` should be replaced by `raw` for all three links?

**Comment by dhewg** [comment]
d'oh, thanks!

File: package/kernel/lantiq/ltq-vdsl-vr11-mei/patches/001-fix-compile.patch [hunk 3]
```diff
@@ -0,0 +1,103 @@
+--- a/src/drv_mei_cpe_linux.h
++++ b/src/drv_mei_cpe_linux.h
+@@ -31,6 +31,9 @@
+ #include <linux/module.h>
+ 
+ #include <linux/sched.h>
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0))
```
Comment (dangowrt) [inline package/kernel/lantiq/ltq-vdsl-vr11-mei/patches/001-fix-compile.patch:7]: I reckon the LINUX_VERSION_CODE #ifdef'ery can be removed here and below, we don't support older kernels.

File: package/kernel/lantiq/ltq-vdsl-vr11/patches/001-fix-compile.patch [hunk 4]
```diff
@@ -0,0 +1,89 @@
+--- a/src/include/drv_dsl_cpe_os_linux.h
++++ b/src/include/drv_dsl_cpe_os_linux.h
+@@ -36,6 +36,9 @@
+ #endif
+ 
+ #include <linux/sched.h>
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0))
```
Comment (dangowrt) [inline package/kernel/lantiq/ltq-vdsl-vr11/patches/001-fix-compile.patch:7]: Also here please drop #ifdef'ery to support ancient kernels

File: package/kernel/lantiq/ltq-vdsl-vr11-mei/patches/001-fix-compile.patch (hunk 3 above)
Comment (dhewg) [reply package/kernel/lantiq/ltq-vdsl-vr11-mei/patches/001-fix-compile.patch]: Done, handling git-less patches with incremental conflicts is way too painful though, please don't make me do that again :)

**Comment by dangowrt** [comment]
Picked to master branch. Thank you!

**Comment by dhewg** [comment]
Nice, thanks everybody!

This is a huge improvement for end users when choosing devices with working DSL, the prior supported ones are all lantiq based. I.e. worth a mention in future release notes.

**Comment by dangowrt** [comment]
> worth a mention in future release notes.

Definitely, that will also draw more attention to the firmware licensing issue, maybe Intel (or Maxlinear by now?) will have mercy and put those blobs under a license which allows redistributing them unconditionally.
